scripts/tests/twister_blackbox/test_runner.py

```diff
@@ -0,0 +1,986 @@
+            ['qemu_x86'],
+            ['device'],
+            ['dummy.agnostic.group2 SKIPPED: Command line testsuite tag filter',
+             'dummy.agnostic.group1.subgroup2 SKIPPED: Command line testsuite tag filter',
+             'dummy.agnostic.group1.subgroup1 SKIPPED: Command line testsuite tag filter',
+             r'0 of 4 test configurations passed \(0.00%\), 0 failed, 0 errored, 4 skipped'
+             ]
+        ),
+        (
+            os.path.join(TEST_DATA, 'tests', 'dummy'),
+            ['qemu_x86'],
+            ['subgrouped'],
+            ['dummy.agnostic.group2 SKIPPED: Command line testsuite tag filter',
+             r'2 of 4 test configurations passed \(100.00%\), 0 failed, 0 errored, 2 skipped'
+             ]
+        ),
+        (
+            os.path.join(TEST_DATA, 'tests', 'dummy'),
+            ['qemu_x86'],
+            ['agnostic', 'device'],
+            [r'3 of 4 test configurations passed \(100.00%\), 0 failed, 0 errored, 1 skipped']
+        ),
+    ]
+    TESTDATA_13 = [
+        (
+            os.path.join(TEST_DATA, 'tests', 'one_fail_one_pass'),
+            ['qemu_x86'],
+            {
+                'selected_test_instances': 2,
+                'skipped_configurations': 0,
+                'passed_configurations': 0,
+                'failed_configurations': 1,
+                'errored_configurations': 0,
+            }
+        )
+    ]
+
+    TESTDATA_14 = [
+        (
+            os.path.join(TEST_DATA, 'tests', 'always_build_error'),
+            ['qemu_x86_64'],
+            '1',
+        ),
+        (
+            os.path.join(TEST_DATA, 'tests', 'always_build_error'),
+            ['qemu_x86'],
+            '4',
+        ),
+    ]
+
+    @classmethod
+    def setup_class(cls):
+        apath = os.path.join(ZEPHYR_BASE, 'scripts', 'twister')
+        cls.loader = importlib.machinery.SourceFileLoader('__main__', apath)
+        cls.spec = importlib.util.spec_from_loader(cls.loader.name, cls.loader)
+        cls.twister_module = importlib.util.module_from_spec(cls.spec)
+
+    @classmethod
+    def teardown_class(cls):
+        pass
+
+    @pytest.mark.usefixtures("clear_log")
+    @pytest.mark.parametrize(
+        'test_path, test_platforms, expected',
+        TESTDATA_1,
+        ids=[
+            'emulation_only tests/dummy/agnostic',
+            'emulation_only tests/dummy/device',
+        ]
+    )
+    def test_emulation_only(self, capfd, test_path, test_platforms, expected):
+        args = ['-i', '-T', test_path, '--emulation-only'] + \
+               [val for pair in zip(
+                   ['-p'] * len(test_platforms), test_platforms
+               ) for val in pair]
+
+        with mock.patch.object(sys, 'argv', [sys.argv[0]] + args), \
+            pytest.raises(SystemExit) as sys_exit:
+            self.loader.exec_module(self.twister_module)
+
+        select_regex = r'^INFO    - (?P<test_scenarios>[0-9]+) test scenarios' \
```

Comment:
this string match is tight coupled with code output which may changed, and do we expect to update this every time we change the output?

Reply:
> this string match is tight coupled with code output which may changed, and do we expect to update this every time we change the output?

All developers should fix any tests they break and create new ones when introducing new features. If we assume that everything can change, why test?

Reply:
test is for either:
1. old function is not break.
2. new funtion is valid.
but now this coding style will make 
1.  always change old case to fit the new style. so the regression become meanless.
2.  match exact out is not a function, but too tight couple with code under test. it is not test functionality, it is testing the output. this is not a fucntion test for script but a fixed IO testing, which means the output shall not be change, and if they do, we need change the testcase, but at the saem time, we do not lock the output, but leave to open to change, so why add such case?

Reply:
`twister.log` and `std` is certainly the most volatile of our information sources in blackbox tests and thus should be used sparingly to verify the tests' correctness.

I will verify whether I can switch my tests to more stable information sources, like `testplan.json` or similar. However, it is highly probable that at least some of the tests will necessitate checking `twister.log`, `build.log` or `stdout`, etc.

If developer's change breaks tests, we expect of them to fix their code or modify the test to account for the change.

Here, we have a tight coupling with a strict regex, because we want to extract numbers of tests run/build/failed/errored/etc. and `stderr` has been a natural place to find that information, as that's where it's output to the Twister user. We do not expect the output format to change in the future.

Reply:
> If developer's change breaks tests, we expect of them to fix their code or modify the test to account for the change.

if they change to test case to meet new changes, then how about those who rely on those old interfaces and outputs? with this test, it seems to protect this interface and output consistant, however it turns out to be not protect anything, in this case, why test?